hivemind/client/averaging/__init__.py

```diff
@@ -324,35 +312,43 @@ async def _step(self, *, future: MPFuture, gather_binary: bytes, weight: float,
                 future.set_exception(RuntimeError("Internal sanity check failed: averager.step left future pending."
                                                   " Please report this to hivemind issues."))
 
-    async def _make_allreduce_runner(self, group_info: GroupInfo, min_vector_size: int, **kwargs) -> AllReduceRunner:
-        """ Use a group description found by Matchmaking to form AllreduceRunner """
+    async def _run_allreduce(self, group_info: GroupInfo, min_vector_size: int, **kwargs) -> GatheredData:
+        """ run all-reduce in a given group and update tensors in-place, return gathered metadata """
         try:
```

Comment:
Old parameterization: you define one all-reduce, we run it.

New parameterization: we give you group ids, you do whatever TF you want with them

For instance, PowerSGD can run 2 allreduce runners in a sequence